server/api/batch_inventory.py

```diff
@@ -0,0 +1,466 @@
+
+        for choice in contest.choices:
+            batch_tallies[batch_key][choice.name] += int(
+                column_value(row, choice.name, cvr_number, choice_indices)
+            )
+
+    election_results: ElectionResults = dict(
+        ballot_count_by_batch=dict_to_items_list(ballot_count_by_batch),
+        ballot_count_by_group=dict(ballot_count_by_group),
+        batch_tallies=dict_to_items_list(batch_tallies),
+    )
+    batch_inventory_data.election_results = election_results
+
+    # If tabulator status file already uploaded, try reprocessing it, since it
+    # validates tabulator names against the CVR file.
+    if batch_inventory_data.tabulator_status_file:
+        batch_inventory_data.tabulator_id_to_name = None
+        batch_inventory_data.tabulator_status_file.task = create_background_task(
+            process_batch_inventory_tabulator_status_file,
+            dict(jurisdiction_id=jurisdiction.id),
+        )
+
+
+@background_task
+def process_batch_inventory_tabulator_status_file(jurisdiction_id: str):
+    batch_inventory_data = BatchInventoryData.query.get(jurisdiction_id)
+    file = retrieve_file(batch_inventory_data.tabulator_status_file.storage_path)
+    cvr_xml = ElementTree.parse(file)
+
+    tabulators = cvr_xml.findall("tabulators/tb")
+    tabulator_id_to_name = {
+        tabulator.get("tid"): tabulator.get("name") for tabulator in tabulators
+    }
+
+    ballot_count_by_batch = items_list_to_dict(
+        batch_inventory_data.election_results["ballot_count_by_batch"]
+    )
+    cvr_tabulator_ids = {
+        tabulator_id for (tabulator_id, _) in ballot_count_by_batch.keys()
+    }
+    missing_tabulators = set(cvr_tabulator_ids) - set(tabulator_id_to_name.keys())
+    if len(missing_tabulators) > 0:
+        raise UserError(
+            "Could not find some tabulators from CVR file in Tabulator Status file."
+            f" Missing tabulator IDs: {', '.join(missing_tabulators)}."
+        )
+
+    batch_inventory_data.tabulator_id_to_name = tabulator_id_to_name
+
+
+@api.route(
+    "/election/<election_id>/jurisdiction/<jurisdiction_id>/batch-inventory/cvr",
+    methods=["PUT"],
+)
+@restrict_access([UserType.JURISDICTION_ADMIN])
+def upload_batch_inventory_cvr(election: Election, jurisdiction: Jurisdiction):
+    if len(list(jurisdiction.contests)) == 0:
+        raise Conflict("Jurisdiction does not have any contests assigned")
+
+    validate_csv_mimetype(request.files["cvr"])
+    file_name = request.files["cvr"].filename
+    storage_path = store_file(
+        request.files["cvr"].stream,
+        f"audits/{election.id}/jurisdictions/{jurisdiction.id}/"
+        + timestamp_filename("batch-inventory-cvrs", "csv"),
+    )
+
+    batch_inventory_data = BatchInventoryData.query.get(jurisdiction.id)
+    if not batch_inventory_data:
+        batch_inventory_data = BatchInventoryData(jurisdiction_id=jurisdiction.id)
+        db_session.add(batch_inventory_data)
```

Comment:
One of the downsides to having a separate table is that we have to do some gymnastics to create the BatchInventoryData object only when it's needed. I thought it was worthwhile, but open to thoughts on this

Reply:
I see below that `upload_batch_inventory_tabulator_status` does not have a case for `if not batch_inventory_data`. Is the assumption that the client will always call `upload_batch_inventory_cvr` first?

Wondering if it's safer to abstract this `if not batch_inventory_data` check to a helper that both call or a decorator that both use

Reply:
Yeah, the frontend requires the files be uploaded in order. I agree that ideally the backend wouldn't assume that without enforcing it. Maybe I'll just add a quick line of validation to enforce that.

Reply:
Looping back to this comment: Were you still planning to add that validation to `upload_batch_inventory_tabulator_status`?

Reply:
Oops, forgot that, thanks! 58228a3f